File: atcoder/abc178C_ubiquity.lua
-- Vicfred
-- https://atcoder.jp/contests/abc178/tasks/abc178_c
-- math

function binpow(a, b, m)
    a = a % m
    res = 1
    while b > 0 do
        if (b % 2) == 1 then
            res = (res * a) % m
        end
        a = (a * a) % m
        b = b // 2
    end

    return res
end

mod = 1000000007

n = io.read("*n")

ans = binpow(10, n, mod) - binpow(9, n, mod) - binpow(9, n, mod) + binpow(8, n, mod)
ans = ans + mod
ans = ans % mod
print(ans)

File: atcoder/abc178C_ubiquity_.hs
-- uninhm
-- https://atcoder.jp/contests/abc178/tasks/abc178_c
-- math, combinatorics

mmod = 10^9+7

binpow a b
  | b == 0    = 1
  | otherwise =
    if b `mod` 2 == 1 then
      res `mod` mmod * res `mod` mmod * a `mod` mmod
    else
      res `mod` mmod * res `mod` mmod
        where res = binpow a (b `div` 2)

solve n = (binpow 10 n) - ((*2) $ binpow 9 n) + (binpow 8 n)
normalize x = (x + mmod) `mod` mmod

main = interact $ show . normalize . solve . read
